noir-projects/aztec-nr/aztec/src/keys/getters.nr

```diff
@@ -19,8 +19,28 @@ pub fn get_npk_m(context: &mut PrivateContext, address: AztecAddress) -> Grumpki
     get_master_key(context, address, NULLIFIER_INDEX)
 }
 
+// For historical access, we move the fetching of the header to the top level to minimize redundant fetching / proving validity of the header.
+pub fn get_npk_m_at(
+    context: &mut PrivateContext,
+    address: AztecAddress,
+    block_number: u32,
+    header: Header
+) -> GrumpkinPoint {
+    assert_eq(block_number, header.global_variables.block_number as u32);
```

Comment:
This feels strange, why do you want to pass the `block_number` if it is the same value as we have in the header?

Just a small thing that popped up in my mind. As we seem to generally want to supply multiple ways to execute these functions on the header, should we take a look at having them as a trait on the header? 

In that way, you would get the header that you are interested in for data, and then you can use the "getters". 

```rust
let header = context.get_header_at(block_number);

let npk_m = header.get_npk_m(context, address);

// and for historical access to public state
let storage = header.public_storage_historical_read(address, storage_slot);
```

@nventuro what are you thoughts? Should allow us to get rid of some code and think it is similar or better ux, but your call 🫡 

If to change like this, might be better to handle in separate PR since we have the flow a few places that was otherwise untouched here. Think this could essentially get rid of almost all of the `_at` function we got for time, just keep the `get_header_at`.